### ISSUE TITLE
Few suggestions for refund functionality

### ISSUE THREAD

**Issue by knit-pay** [wp-pay/core#50]
It was a great initiative to add a refund option in the wp-pay core library. I have few suggestions for the same.

1. I can see that currently error/exception is not handled in case any issue occurs while creating a refund (in `$gateway->create_refund`). are you planning to handle errors after this function?
https://github.com/wp-pay/core/blob/c69d05c3507d5ca15f7a85d380f721e9a7ff460e/src/Plugin.php#L1093

2. In Mollie gateway, I can see that the payment object is fetched from the transaction id. If we will see other functions like `start` and `update_status`, $payment is passed as a parameter and is getting used directly. I think in almost all the payment gateways, a payment object will be needed for initiating the refund. Wouldn't it be great if we will fetch the payment object in the Plugin class and pass it as a parameter, instead of transaction id. It will reduce rework in all the other gateways. What do you think about it?
https://github.com/wp-pay-gateways/mollie/blob/d58e58f1d0d748f33baa5486f9741f47a2a64c77/src/Gateway.php#L851

3. Refunded amount is getting update in `update_status` function. Shouldn't we update it while creating refund in `create_refund` function?
https://github.com/wp-pay-gateways/mollie/blob/d58e58f1d0d748f33baa5486f9741f47a2a64c77/src/Gateway.php#L770

4. Should the status of payment be marked as "REFUNDED" in `create_refund` function.

**Comment by remcotolsma** [comment]
> 1. I can see that currently error/exception is not handled in case any issue occurs while creating a refund (in $gateway->create_refund). are you planning to handle errors after this function?

Yes, see for an example: https://github.com/wp-pay-extensions/easy-digital-downloads/blob/fb3f184913f256ee33b925faf331dca34735f0c4/src/RefundsManager.php#L76-L103.

> 2. In Mollie gateway, I can see that the payment object is fetched from the transaction id. If we will see other functions like start and update_status, $payment is passed as a parameter and is getting used directly.

Great idea, i already opened an issue about the `Plugin::create_refund( $transaction_id, $gateway, Money $amount, $description = null )` signature here: https://github.com/pronamic/wp-pronamic-pay/issues/131. There must be a way to also pass on a partial amount or a credit invoice number, etc.

@rvdsteege need arises  😉

> 3. Refunded amount is getting update in update_status function. Shouldn't we update it while creating refund in create_refund function?

Yes, but many payment providers also offer the option to perform a refund from their dashboard. That is why we also synchronize the refunded amount in the update status actions. 

> 4. Should the status of payment be marked as "REFUNDED" in create_refund function.

We are considering dropping the 'refunded' status and deriving this from the refunded amount.

> 5. Are you planning to handle both Partial and Full refunds?

Yes, we already extend Easy Digital Downloads and Gravity Forms with support for partial refunds:

**Easy Digital Downloads**

- https://github.com/wp-pay-extensions/easy-digital-downloads/blob/3258e4283889d3f7f3497251439a1cd33ade9063/src/Extension.php#L516-L518
- https://github.com/wp-pay-extensions/easy-digital-downloads/blob/fb3f184913f256ee33b925faf331dca34735f0c4/src/RefundsManager.php#L179

**Gravity Forms**

- https://github.com/wp-pay-extensions/gravityforms/blob/e7f406c4ecda316df2f8fc8a73037e55784dfcbc/src/Extension.php#L1640-L1662
- https://github.com/wp-pay-extensions/gravityforms/blob/e7f406c4ecda316df2f8fc8a73037e55784dfcbc/src/Extension.php#L730-L735

**Comment by knit-pay** [comment]
Thanks for the information.
